sample.go

```diff
@@ -456,31 +435,78 @@ func (s *SampleSnapshot) ResetCount() {
 }
 
 // Max returns the maximal value at the time the snapshot was taken.
-func (s *SampleSnapshot) Max() int64 { return SampleMax(s.values) }
+func (s *SampleSnapshot) Max() int64 {
+	if len(s.values) == 0 {
+		return 0
+	}
+	return s.values[len(s.values)-1]
+}
 
 // Mean returns the mean value at the time the snapshot was taken.
 func (s *SampleSnapshot) Mean() float64 { return SampleMean(s.values) }
 
 // Min returns the minimal value at the time the snapshot was taken.
-func (s *SampleSnapshot) Min() int64 { return SampleMin(s.values) }
+func (s *SampleSnapshot) Min() int64 {
+	if len(s.values) == 0 {
+		return 0
+	}
+	return s.values[0]
+}
 
 // Percentile returns an arbitrary percentile of values at the time the
 // snapshot was taken.
 func (s *SampleSnapshot) Percentile(p float64) float64 {
-	return SamplePercentile(s.values, p)
+	return s.Percentiles([]float64{p})[0]
 }
 
 // Percentiles returns a slice of arbitrary percentiles of values at the time
 // the snapshot was taken.
 func (s *SampleSnapshot) Percentiles(ps []float64) []float64 {
-	return SamplePercentiles(s.values, ps)
+	scores := make([]float64, len(ps))
+	size := len(s.values)
+	if size > 0 {
+		for i, p := range ps {
+			pos := p * float64(size+1)
+			if pos < 1.0 {
+				scores[i] = float64(s.values[0])
+			} else if pos >= float64(size) {
+				scores[i] = float64(s.values[size-1])
+			} else {
+				lower := float64(s.values[int(pos)-1])
+				upper := float64(s.values[int(pos)])
+				scores[i] = lower + (pos-math.Floor(pos))*(upper-lower)
+			}
+		}
+	}
+	return scores
+}
+
+// Rank returns the rank of a value against the sample
+func (s *SampleSnapshot) Rank(in int64) float64 {
+	// return -1 if there are no values
+	if len(s.values) == 0 {
+		return -1
+	}
+
+	// return the index of the value or where it should be
+	idx := float64(sort.Search(len(s.values), func(i int) bool { return s.values[i] > in }))
+
+	// add 1 to the length of val since we aren't inserting the value
+	return float64(idx) / float64(len(s.values)+1)
```

Comment:
i've been meaning to talk to joe and do some interpolation here like the above percentiles method does... please do since we're in this area.

Reply:
Also, why the +1? that's from a pretty old version of this code. the current master version does "	return float64(i) / (float64(s.list.Size()))"